drivers: kconfig: revise the default alignment to 4-bytes

attie-argentum opened the pull request:
Since a85ffa8, the sd_card.card_buffer has been at an unaligned offset, meaning that for systems using the default alignment value of 1, a hard fault will present when executing code that casts the buffer to uint32_t, such as sdmmc_spi_read_cxd(), card_query_written(), and possibly others...

Historically, it appears that the alignment of card_buffer was good and operational only by chance.

See #62619 for further details.

nordicjm commented:
I think you are fixing the wrong thing here, alignment of 4 is _not_ required for all devices, the reason it is required for your particular board is because you are using a Cortex M0, which does not support unaligned memory access. The fix should be to add the `__packed` attribute to the structs that need this

attie-argentum commented:
Hi @nordicjm - please could you clarify where you'd like to see `__packed`? I'm not sure how that would help in this situation...

attie-argentum commented:
I've just spotted that `SD_BUFFER_SIZE` will grow significantly in this case (if `MMC_STACK=n`)...

https://github.com/zephyrproject-rtos/zephyr/blob/38c918164217965c92b04f23ca894f926bc11219/subsys/sd/Kconfig#L68-L79

danieldegrasse commented:
> I've just spotted that SD_BUFFER_SIZE will grow significantly in this case (if MMC_STACK=n)...

This is a good catch- the issue here is that `CONFIG_SDHC_BUFFER_ALIGNMENT` is primarily intended to aid in DMA operations. Maybe we should instead adjust the declaration of the card buffer structure: https://github.com/zephyrproject-rtos/zephyr/blob/38c918164217965c92b04f23ca894f926bc11219/include/zephyr/sd/sd.h#L43-L70

to use something like `MAX(4, CONFIG_SDHC_BUFFER_ALIGNMENT)` to determine the card buffer alignment, as you initially suggested @attie-argentum?

attie-argentum commented:
> Maybe we should instead adjust the declaration of the card buffer structure [...] to use something like `MAX(4, CONFIG_SDHC_BUFFER_ALIGNMENT)`

This would work, but may lead to unexpected situations, as sometimes the alignment won't be `CONFIG_SDHC_BUFFER_ALIGNMENT` bytes... but is that a problem? @nordicjm what do you think?

---

This also feels a bit tangled / confused - why does buffer alignment infer requiring a whole block?

https://github.com/zephyrproject-rtos/zephyr/blob/38c918164217965c92b04f23ca894f926bc11219/subsys/sd/Kconfig#L70-L72

`SDHC_BUFFER_ALIGNMENT` only appears to be set different to the default in the two following to locations - perhaps we should also be setting `SD_BUFFER_SIZE=512` here too? ... Apologies if my ignorance is showing through!

https://github.com/zephyrproject-rtos/zephyr/blob/38c918164217965c92b04f23ca894f926bc11219/drivers/sdhc/Kconfig.imx#L31-L33

https://github.com/zephyrproject-rtos/zephyr/blob/38c918164217965c92b04f23ca894f926bc11219/drivers/sdhc/Kconfig.mcux_sdif#L24-L26

danieldegrasse commented:
> SDHC_BUFFER_ALIGNMENT only appears to be set different to the default in the two following to locations - perhaps we should also be setting SD_BUFFER_SIZE=512 here too? ... apologies if my ignorance is showing through!

No worries- the reason we set `CONFIG_SD_BUFFER_SIZE` to a block there is because when `CONFIG_SDHC_BUFFER_ALIGNMENT != 1`, that denotes that the host controller expects all data buffers to be aligned to a given boundary. In this case, we may need up to a block size worth of data in the SD card buffer, since we will need to copy any unaligned data buffers into the card buffer before passing that buffer to the SD host controller.

The specific issue you are encountering with unaligned data access on cortex M0 is different from what `CONFIG_SDHC_BUFFER_ALIGNMENT` is intended to solve, since there should not be a need for the intermediate buffer here (since I assume the compiler will already have aligned data buffers being passed to the sd card subsystem). I suspect that when `CONFIG_SDHC_BUFFER_ALIGNMENT=1` the __aligned(1) statement in the `struct sd_card` is what results in this buffer being unaligned when GCC would otherwise align it properly. Therefore I think the best fix may be to only enforce alignment when `CONFIG_SDHC_BUFFER_ALIGNMENT !=1`. We could change the buffer declaration to something like this:
```
#if CONFIG_SDHC_BUFFER_ALIGNMENT != 1
uint8_t card_buffer[CONFIG_SD_BUFFER_SIZE]
		__aligned(CONFIG_SDHC_BUFFER_ALIGNMENT); /* Card internal buffer */
#else
uint8_t card_buffer[CONFIG_SD_BUFFER_SIZE]
#endif
```

Then the SD buffer will not be enlarged and the code to copy unaligned buffers won't run, but the buffer itself will still be aligned to prevent this sort of issue (since GCC should use the best alignment for the platform when there is no `__aligned` statement)

attie-argentum commented:
Ping @danieldegrasse

danieldegrasse commented:
> Unfortunately removing the __aligned tag doesn't work - the member is still misaligned.

@nordicjm- do you have any thoughts here? I would expect that GCC would enforce alignment to avoid unaligned memory access on structures like this when compiling for cortex M0 CPUs, if an alignment is not set manually with `__aligned`.

> Shall we proceed with __aligned(MAX(4, CONFIG_SDHC_BUFFER_ALIGNMENT))

I think this is an OK approach. I'm not aware of any architectures that don't support unaligned access and have alignment restrictions more aggressive than 4 bytes

nordicjm commented:
> > Unfortunately removing the __aligned tag doesn't work - the member is still misaligned.
> 
> @nordicjm- do you have any thoughts here? I would expect that GCC would enforce alignment to avoid unaligned memory access on structures like this when compiling for cortex M0 CPUs, if an alignment is not set manually with `__aligned`.
> 
> > Shall we proceed with __aligned(MAX(4, CONFIG_SDHC_BUFFER_ALIGNMENT))
> 
> I think this is an OK approach. I'm not aware of any architectures that don't support unaligned access and have alignment restrictions more aggressive than 4 bytes

As said before, you need to use `__packed`

attie-argentum commented:
> > > The fix should be to add the `__packed` attribute to the structs that need this
>>
> > please could you clarify where you'd like to see `__packed`?
>
> As said before, you need to use `__packed`

@nordicjm - please could you clarify where you'd like to see `__packed`?

The following doesn't resolve the issue, and I wouldn't really expect it to... this will close the gaps between each member to reduce the overall struct's size (the opposite of what we're after for `sd_card.card_buffer`).

```patch
 struct sd_card {
     const struct device *sdhc; /*!< SD host controller for card */
     struct sdhc_io bus_io; /*!< Current bus I/O props for SDHC */
     enum sd_voltage card_voltage; /*!< Card signal voltage */
     struct k_mutex lock; /*!< card mutex */
     struct sdhc_host_props host_props; /*!< SDHC host properties */
     uint32_t ocr; /*!< Raw card OCR content */
     struct sd_switch_caps switch_caps; /*!< SD switch capabilities */
     unsigned int num_io: 3; /*!< I/O function count. 0 for SD cards */
     uint16_t relative_addr; /*!< Card relative address */
     uint32_t block_count; /*!< Number of blocks in SD card */
     uint16_t block_size; /*!< SD block size */
     uint8_t sd_version; /*!< SD specification version */
     uint8_t card_speed; /*!< Card timing mode */
     enum card_status status; /*!< Card status */
     enum card_type type; /*!< Card type */
     uint16_t flags; /*!< Card flags */
     uint8_t bus_width; /*!< Desired bus width */
     uint8_t card_buffer[CONFIG_SD_BUFFER_SIZE]
         __aligned(CONFIG_SDHC_BUFFER_ALIGNMENT); /* Card internal buffer */
-};
+} __packed;
```

```
*** Booting Zephyr OS build zephyr-v3.4.0-4640-g815e7a1e3434 ***
[00:00:00.041,000] <err> os: ***** HARD FAULT *****
[00:00:00.041,000] <err> os: r0/a1:  0x00000000  r1/a2:  0x00000000  r2/a3:  0x00000000
[00:00:00.041,000] <err> os: r3/a4:  0x00000001 r12/ip:  0x00000003 r14/lr:  0x00006617
[00:00:00.041,000] <err> os:  xpsr:  0x01000000
[00:00:00.041,000] <err> os: Faulting instruction address (r15/pc): 0x00008098
[00:00:00.041,000] <err> os: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0
[00:00:00.041,000] <err> os: Current thread: 0x200007e8 (unknown)
[00:00:00.092,000] <err> os: Halting system
```

The output confirms my concern - these are all new warnings:

```
$ ninja
[5/13] Building C object zephyr/subsys/sd/CMakeFiles/subsys__sd.dir/sd.c.obj
/p/zephyr/subsys/sd/sd.c: In function 'sd_init_io':
/p/zephyr/subsys/sd/sd.c:119:34: warning: taking address of packed member of 'struct sd_card' may result in an unaligned pointer value [-Waddress-of-packed-member]
  119 |         struct sdhc_io *bus_io = &card->bus_io;
      |                                  ^~~~~~~~~~~~~
/p/zephyr/subsys/sd/sd.c: In function 'sd_init':
/p/zephyr/subsys/sd/sd.c:223:47: warning: taking address of packed member of 'struct sd_card' may result in an unaligned pointer value [-Waddress-of-packed-member]
  223 |         ret = sdhc_get_host_props(card->sdhc, &card->host_props);
      |                                               ^~~~~~~~~~~~~~~~~
/p/zephyr/subsys/sd/sd.c:230:28: warning: taking address of packed member of 'struct sd_card' may result in an unaligned pointer value [-Waddress-of-packed-member]
  230 |         ret = k_mutex_init(&card->lock);
      |                            ^~~~~~~~~~~
/p/zephyr/subsys/sd/sd.c:235:28: warning: taking address of packed member of 'struct sd_card' may result in an unaligned pointer value [-Waddress-of-packed-member]
  235 |         ret = k_mutex_lock(&card->lock, K_MSEC(CONFIG_SD_INIT_TIMEOUT));
      |                            ^~~~~~~~~~~
[...]
```

attie-argentum committed:
include: sd: revise the minimum alignment of card_buffer to 4-bytes

Since a85ffa8, `struct sd_card.card_buffer` has naturally fallen at an
offset that is unaligned. This meant that on systems without support for
unaligned access (e.g: Cortex-M0), a hard fault would present when
executing code that casts the buffer to `uint32_t` (such as
`sdmmc_spi_read_cxd()`, `card_query_written()`, etc...)

Historically, it appears that the alignment of the `card_buffer` member
was good and operational only by chance.

Altering the default value of `CONFIG_SDHC_BUFFER_ALIGNMENT` was
rejected, as this has wider implications.

Fixes #62619

Signed-off-by: Attie Grande <attie.grande@argentum-systems.co.uk>

attie-argentum commented:
> > Shall we proceed with __aligned(MAX(4, CONFIG_SDHC_BUFFER_ALIGNMENT))
>
> I think this is an OK approach. I'm not aware of any architectures that don't support unaligned access and have alignment restrictions more aggressive than 4 bytes

@danieldegrasse - great, thanks! I've reworked the PR.

danieldegrasse commented:
@attie-argentum could you revise this line in the commit message:
> See https://github.com/zephyrproject-rtos/zephyr/issues/62619 and https://github.com/zephyrproject-rtos/zephyr/pull/62669 for further details.

To the following:
> Fixes #62619

This should link the PR and the issue, and will make it possible to merge this PR during the code freeze window (since it is fixing a bug)

attie-argentum commented:
> could you revise this line in the commit message

Done, I've had push-back on that so started avoiding it.

danieldegrasse commented:
@decsny, could you take a look at this?

danieldegrasse commented:
@attie-argentum is there a specific platform/sha you have reproduced this on? I'd like to see if I can reproduce the issue to better understand why the alignment is off.

danieldegrasse commented:
> I put my notes in the original issue, https://github.com/zephyrproject-rtos/zephyr/issues/62619#issuecomment-1718325272 (using a SAML21, but I believe any Cortex-M0(+) should reproduce the issue).

Sorry for the delay- took some time to revisit this, and I think the approach you've picked is the best option given the way the SD subsystem manages the card buffer. Perhaps add a comment above the buffer declaration stating `/* Buffer is accessed as as (uint32_t *) by SD subsystem, so it must be aligned to 4 bytes for platforms that do not support unaligned access */`

@decsny, could you revisit? The core issue is that we declare the `card_buffer` as a uint8_t array. The compiler therefore assumes there is no need to align the buffer, and will place it aligned to a byte boundary:

```
Reading symbols from build/zephyr/zephyr.elf...
(gdb) p &sdmmc_data_0.card.card_buffer
$1 = (uint8_t (*)[64]) 0x200000b5 <sdmmc_data_0+113>
(gdb)
```

If we declared the buffer as a uint32_t array, this would also resolve the alignment issue, but we'd need to cast the buffer to a `uint8_t*` within many functions in the SD subsystem. Based on this, I think enforcing the alignment restriction here might be the best fix.

attie-argentum commented:
Any chance we can progress this PR? (ping @decsny)
